sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/ParquetNestedPredicatePushDownBenchmark.scala

```diff
@@ -0,0 +1,120 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.benchmark
+
+import org.apache.spark.SparkConf
+import org.apache.spark.benchmark.Benchmark
+import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Synthetic benchmark for nested fields predicate push down performance for Parquet datasource.
+ * To run this benchmark:
+ * {{{
+ *   1. without sbt:
+ *      bin/spark-submit --class <this class> --jars <spark core test jar> <sql core test jar>
+ *   2. build/sbt "sql/test:runMain <this class>"
+ *   3. generate result:
+ *      SPARK_GENERATE_BENCHMARK_FILES=1 build/sbt "sql/test:runMain <this class>"
+ *      Results will be written to "benchmarks/ParquetNestedPredicatePushDownBenchmark-results.txt".
+ * }}}
+ */
+object ParquetNestedPredicatePushDownBenchmark extends SqlBasedBenchmark {
+
+  private val N = 100 * 1024 * 1024
+  private val NUMBER_OF_ITER = 10
+
+  override def getSparkSession: SparkSession = {
+    val conf = new SparkConf()
+      .setAppName(this.getClass.getSimpleName)
+      // Since `spark.master` always exists, overrides this value
+      .set("spark.master", "local[1]")
+
+    SparkSession.builder().config(conf).getOrCreate()
+  }
+
+  private val df: DataFrame = spark.range(1, N, 1, 4).toDF("id")
+    .selectExpr("id", "STRUCT(id x, STRUCT(CAST(id AS STRING) z) y) nested")
+    .sort("id")
+
+  private def addCase(benchmark: Benchmark, inputPath: String,
+                      enableNestedPD: Boolean,
+                      name: String, filterFn: DataFrame => DataFrame): Unit = {
+    val loadDF = spark.read.parquet(inputPath)
+    benchmark.addCase(name) {
+      _ =>
+        withSQLConf((SQLConf.NESTED_PREDICATE_PUSHDOWN_ENABLED.key,
+          enableNestedPD.toString)) {
+          filterFn(loadDF).noop()
+        }
+    }
+  }
+
+  private def createAndRunBenchmark(name: String, filterFn: DataFrame => DataFrame): Unit = {
+    withTempPath {
+      tempDir =>
+        val outputPath = tempDir.getCanonicalPath
+        df.write.mode(SaveMode.Overwrite).parquet(tempDir.getCanonicalPath)
+        val benchmark = new Benchmark(name, N, NUMBER_OF_ITER, output = output)
+        addCase(benchmark, outputPath, enableNestedPD = false,
+          "NestedFieldsPredicatePushDownDisabled", filterFn)
```

Comment:
Maybe just call it, `With Nested Predicate Pushdown` and `Without Nested Predicate Pushdown` for readability?